### ISSUE TITLE
Clothes layering displaying weird behaviour

### ISSUE THREAD

**Issue by Poupine** [hbertiche/PBNS#2]
Hi,

I've been looking at PBNS for clothes resizing and I've hit a problem that I can't quite explain with clothes layering. On some of the clothes I am working with, the artist modeled volume around arm seams by bending the surface of the fabric inward. I thus set these vertices as inner layers when training and was expecting these vertices to end up as inner layers.
Unfortunately, I end up with seems bend over backward (they are now on the outside instead of on the inside) and I can't quite rack my head around the reason why that would happen.
Here is an example of what I mean: 
![image](https://user-images.githubusercontent.com/29566363/158202016-4b88849d-f245-4853-b6d5-5b8fa581c9e6.png)
Basically, this piece of the surface folds is originally folded the other way around and the closest to the body. I thus set these vertices as inner layers but end up with the image you see up there.

Hope you can help!

PS: Congrats on a great piece of tech ;)

**Comment by Poupine** [comment]
I think I may just have found the reason why this thing was happening: in the collision_loss, line 65-73
``` python
        # add layer to "body"
        if i + 1 < len(layers):
            if stop_gradient:
                v = tf.stop_gradient(v)
                v_vn = tf.stop_gradient(tf.gather(V_vn, l, axis=1))
            else:
                v_vn = tf.gather(V_vn, l, axis=1)
            B = tf.concat((B, v), axis=1)
            B_vn = tf.concat((B_vn, -v_vn), axis=1)
```

When adding the "layer to body", surfaces normals are expected to point in the same direction, which in my case is not the case since the piece I'm adding as the inner layer is bent over backwards. Hence the -v_vn in the last line which solves it :)

**Comment by hbertiche** [comment]
Hi,

Sorry for the delayed response,
although I think you already figured out the issue =)

the collision loss proposed in this work for different layers of cloth assumes these layers are going to be disconnected and that vertex normals will always point outwards,
as you observed, it is possible to adapt the code for specific scenarios where vertices belong to the same piece of cloth,

since I believe you could solve your question, I will close this issue,
feel free to open it again if this is not the case
